### PR TITLE
Lebedeva Ksenia Lab4 Var4

### PR DESCRIPTION
**Pull request by ks2k23**
Write a pass that counts amount of times it was called by other functions in the module, attach this as an attribute to function (func.func) operation.

### CODE REVIEW

File: mlir/lib/Transforms/lab4/lebedeva_ksenia/LebedevaCallFuncCounter.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+class LebedevaCallFuncCounter
+    : public PassWrapper<LebedevaCallFuncCounter, OperationPass<ModuleOp>> {
+private:
+  std::map<StringRef, int> сounter;
+
+public:
+  void runOnOperation() override {
+    getOperation()->walk([&](Operation *op) {
+      if (auto callOp = dyn_cast<LLVM::CallOp>(op)) {
```
Comment (aobolensk) [inline mlir/lib/Transforms/lab4/lebedeva_ksenia/LebedevaCallFuncCounter.cpp]: Why not use walker that goes through `LLVM::CallOp` only?

File: mlir/lib/Transforms/lab4/lebedeva_ksenia/LebedevaCallFuncCounter.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        StringRef functionName = callOp.getCallee().value();
+        сounter[functionName]++;
+      }
+    });
+    getOperation()->walk([&](Operation *op) {
+      if (auto functionOp = dyn_cast<LLVM::LLVMFuncOp>(op)) {
```
Comment (aobolensk) [inline mlir/lib/Transforms/lab4/lebedeva_ksenia/LebedevaCallFuncCounter.cpp]: Same here, no need to filter operations by yourself, LLVM can do that for you

File: mlir/test/Transforms/lab4/lebedeva_ksenia/test.mlir [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  llvm.func @_Z9function1i(%arg0: i32 {llvm.noundef}) -> (i1 {llvm.noundef, llvm.zeroext}) attributes {passthrough = ["mustprogress", "noinline", "nounwind", "optnone", ["uwtable", "2"], ["frame-pointer", "all"], ["min-legal-vector-width", "0"], ["no-trapping-math", "true"], ["stack-protector-buffer-size", "8"], ["target-cpu", "x86-64"], ["target-features", "+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87"], ["tune-cpu", "generic"]]} {
+    // CHECK: attributes {"call-count" = 4
```
Comment (aobolensk) [inline mlir/test/Transforms/lab4/lebedeva_ksenia/test.mlir]: Please, match call count attribute along with function name

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Review by aobolensk — APPROVED** [review]

File: mlir/test/Transforms/lab4/lebedeva_ksenia/test.mlir [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    %1 = llvm.alloca %0 x i32 {alignment = 4 : i64} : (i32) -> !llvm.ptr
+    %2 = llvm.alloca %0 x i32 {alignment = 4 : i64} : (i32) -> !llvm.ptr
+    %3 = llvm.alloca %0 x i8 {alignment = 1 : i64} : (i32) -> !llvm.ptr
+    %4 = llvm.alloca %0 x i8 {alignment = 1 : i64} : (i32) -> !llvm.ptr
+    %5 = llvm.alloca %0 x i32 {alignment = 4 : i64} : (i32) -> !llvm.ptr
+    llvm.store %arg0, %1 {alignment = 4 : i64} : i32, !llvm.ptr
+    llvm.store %arg1, %2 {alignment = 4 : i64} : i32, !llvm.ptr
+    %6 = llvm.load %1 {alignment = 4 : i64} : !llvm.ptr -> i32
+    %7 = llvm.call @_Z9function1i(%6) : (i32) -> i1
+    %8 = llvm.zext %7 : i1 to i8
```
Comment (m-ly4) [inline mlir/test/Transforms/lab4/lebedeva_ksenia/test.mlir:73]:
```suggestion
  llvm.func @extern_func(i32) -> i32
  // CHECK: llvm.func @extern_func
  // CHECK-SAME: attributes {"call-count" = 1 : i32
  llvm.func @_Z9function4ii(%arg0: i32 {llvm.noundef}, %arg1: i32 {llvm.noundef}) attributes {passthrough = ["mustprogress", "noinline", "nounwind", "optnone", ["uwtable", "2"], ["frame-pointer", "all"], ["min-legal-vector-width", "0"], ["no-trapping-math", "true"], ["stack-protector-buffer-size", "8"], ["target-cpu", "x86-64"], ["target-features", "+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87"], ["tune-cpu", "generic"]]} {
    // CHECK: llvm.func @_Z9function4ii
    // CHECK-SAME: attributes {"call-count" = 0 : i32
    %0 = llvm.mlir.constant(1 : i32) : i32
    %1 = llvm.alloca %0 x i32 {alignment = 4 : i64} : (i32) -> !llvm.ptr
    %2 = llvm.alloca %0 x i32 {alignment = 4 : i64} : (i32) -> !llvm.ptr
    %3 = llvm.alloca %0 x i8 {alignment = 1 : i64} : (i32) -> !llvm.ptr
    %4 = llvm.alloca %0 x i8 {alignment = 1 : i64} : (i32) -> !llvm.ptr
    %5 = llvm.alloca %0 x i32 {alignment = 4 : i64} : (i32) -> !llvm.ptr
    llvm.store %arg0, %1 {alignment = 4 : i64} : i32, !llvm.ptr
    llvm.store %arg1, %2 {alignment = 4 : i64} : i32, !llvm.ptr
    %6 = llvm.load %1 {alignment = 4 : i64} : !llvm.ptr -> i32
    %7 = llvm.call @_Z9function1i(%6) : (i32) -> i1
    %8 = llvm.zext %7 : i1 to i8
    %9 = llvm.call @extern_func(%6) : (i32) -> i32
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**